### PR TITLE
Use new listener which implements enhanced tcp features

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Use new listener which implements enhanced tcp features

## Motivation and Context
This package provide customizable TCP net.Listener with various
performance-related options:

 * SO_REUSEPORT. This option allows linear scaling server performance
   on multi-CPU servers.
   See https://www.nginx.com/blog/socket-sharding-nginx-release-1-9-1/ for details.
 * TCP_DEFER_ACCEPT. This option expects the server reads from the accepted
   connection before writing to them.
 * TCP_FASTOPEN. See https://lwn.net/Articles/508865/ for details.

## How Has This Been Tested?
Manually

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#034c8e3b-20f8-4b0b-be3c-b56b0e374316) |
<a name="034c8e3b-20f8-4b0b-be3c-b56b0e374316"></a>
### 6289-ca81bc9/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31445
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a53f114219b8:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 41 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 7 seconds
(4/13) Running aws-sdk-php tests ... done in 6 minutes and 56 seconds
(5/13) Running aws-sdk-ruby tests ... done in 40 seconds
(6/13) Running mc tests ... done in 6 minutes and 7 seconds
(7/13) Running minio-dotnet tests ... done in 6 minutes and 5 seconds
(8/13) Running minio-go tests ... FAILED in 7 minutes and 44 seconds
{
  "args": {
    "policy": "{\"expiration\":\"2018-08-26T02:55:55.713Z\",\"conditions\":[[\"eq\",\"$bucket\",\"minio-go-test-kx5cjnj5wul5zo6t\"],[\"eq\",\"$key\",\"gw2jj2aftp6ytafmqnp9sj3fzj41qg\"],[\"eq\",\"$Content-Type\",\"binary/octet-stream\"],[\"eq\",\"$x-amz-meta-96xrczudc3qgurrjiha0j3tv6ioj13\",\"qhjb32f0gaorfnpysp0f4llfoysx2n\"],[\"content-length-range\", 10, 1048576]]}"
  },
  "duration": 374,
  "error": "400 Bad Request",
  "function": "PresignedPostPolicy(policy)",
  "message": "Http request failed",
  "name": "minio-go: testPresignedPostPolicy",
  "status": "FAIL"
}

Executed 7 out of 13 tests successfully.
```

File: cmd/http/listener.go [hunk 1]
```diff
@@ -314,7 +315,7 @@ func newHTTPListener(serverAddrs []string,
 
 	for _, serverAddr := range serverAddrs {
 		var l net.Listener
-		if l, err = net.Listen("tcp", serverAddr); err != nil {
+		if l, err = NewListener("tcp4", serverAddr); err != nil {
```
Comment (balamurugana) [inline cmd/http/listener.go]: You could keep the name aligned with `net` package like unexported `listen()`

File: cmd/http/server.go [hunk 2]
```diff
@@ -191,7 +191,7 @@ func NewServer(addrs []string, handler http.Handler, getCert certs.GetCertificat
 			CipherSuites:             defaultCipherSuites,
 			CurvePreferences:         secureCurves,
 			MinVersion:               tls.VersionTLS12,
-			NextProtos:               []string{"http/1.1", "h2"},
+			NextProtos:               []string{"h2", "http/1.1"},
```
Comment (balamurugana) [inline cmd/http/server.go]: As per godoc, the order doesn't matter to serve HTTP/2 requests i.e. no need to change it.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by balamurugana — DISMISSED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#a8bbdeab-483f-44a8-87da-b5240b2b28f9) |
<a name="a8bbdeab-483f-44a8-87da-b5240b2b28f9"></a>
### 6289-8a6fa5b/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31730
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b5718410e054:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 45 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 6 minutes and 58 seconds
(5/13) Running aws-sdk-ruby tests ... done in 1 minutes and 4 seconds
(6/13) Running mc tests ... done in 6 minutes and 33 seconds
(7/13) Running minio-dotnet tests ... done in 5 minutes and 50 seconds
(8/13) Running minio-go tests ... done in 10 minutes and 5 seconds
(9/13) Running minio-java tests ... done in 20 minutes and 19 seconds
(10/13) Running minio-js tests ... FAILED in 43 seconds
{
  "name": "minio-js",
  "function": "makeBucket(bucketName, region)",
  "args": "bucketName:minio-js-test-90db34df-08ec-4992-99fd-af749524e0a8-region, region:us-east-1",
  "duration": 40275,
  "status": "FAIL",
  "error": "S3Error: We encountered an internal error, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:56:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:117:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:48:26) at DestroyableTransform.<anonymous> (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:135:12) at finishMaybe (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:371:12) at endWritable (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:378:3) at DestroyableTransform.Writable.end (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:356:5) at IncomingMessage.onend (_stream_readable.js:514:10) at endReadableNT (_stream_readable.js:978:12) at _combinedTickCallback (internal/process/next_tick.js:80:11) at process._tickCallback (internal/process/next_tick.js:104:9)"
}

Executed 9 out of 13 tests successfully.
```

**Review by nitisht — APPROVED** [review]

**Comment by harshavardhana** [comment]
ping @balamurugana

**Review by balamurugana — APPROVED** [review]

**Pull request merged**